src/Common/SecurityToken.cpp

```diff
@@ -188,7 +192,212 @@ namespace VeraCrypt
 			throw Pkcs11Exception(status);
 	}
 
-	vector <SecurityTokenKeyfile> SecurityToken::GetAvailableKeyfiles(CK_SLOT_ID* slotIdFilter, const wstring keyfileIdFilter)
+
+	void SecurityTokenImpl::GetSecurityTokenKey(wstring tokenKeyDescriptor, SecurityTokenKey &key, SecurityTokenKeyOperation mode)
+	{
+
+		CK_SLOT_ID slotId;
+
+		if (swscanf (tokenKeyDescriptor.c_str(), L"%lu", &slotId) != 1)
+			throw InvalidSecurityTokenKeyfilePath();
+
+		size_t n = tokenKeyDescriptor.find(L":");
+		if (n == std::string::npos) {
+			throw InvalidSecurityTokenKeyfilePath();
+		}
+		
+		wstring keyId = tokenKeyDescriptor.substr(n+1);
+
+		vector <SecurityTokenKey> keys;
+		if (mode == SecurityTokenKeyOperation::ENCRYPT) {
+		 	keys = SecurityToken::GetAvailablePublicKeys(&slotId, keyId);
+		} else if (mode == SecurityTokenKeyOperation::DECRYPT) {
+			keys = SecurityToken::GetAvailablePrivateKeys(&slotId, keyId);
+		} else {
+			throw ParameterIncorrect(SRC_POS);
+		}
+		if (keys.size() > 1 || keys.size() == 0) {
+			throw Pkcs11Exception (CKR_TOKEN_NOT_RECOGNIZED);
+		}
+		key = keys[0];
+	}
+
+	vector <SecurityTokenKey> SecurityTokenImpl::GetAvailablePrivateKeys(CK_SLOT_ID *slotIdFilter, const wstring keyIdFilter)
+	{
+		bool unrecognizedTokenPresent = false;
+		vector <SecurityTokenKey> keys;
+
+		foreach (const CK_SLOT_ID &slotId, GetTokenSlots())
+		{
+			SecurityTokenInfo token;
+
+			if (slotIdFilter && *slotIdFilter != slotId)
+				continue;
+
+			try
+			{
+				LoginUserIfRequired (slotId);
+				token = GetTokenInfo (slotId);
+			}
+			catch (UserAbort &)
+			{
+				continue;
+			}
+			catch (Pkcs11Exception &e)
+			{
+				if (e.GetErrorCode() == CKR_TOKEN_NOT_RECOGNIZED)
+				{
+					unrecognizedTokenPresent = true;
+					continue;
+				}
+				throw;
+			}
+
+			foreach (const CK_OBJECT_HANDLE &dataHandle, GetObjects (slotId, CKO_PRIVATE_KEY))
+			{
+				SecurityTokenKey key;
+				key.Handle = dataHandle;
+				key.SlotId = slotId;
+				key.Token = token;
+
+				vector <uint8> privateAttrib;
+				GetObjectAttribute (slotId, dataHandle, CKA_PRIVATE, privateAttrib);
+
+				if (privateAttrib.size() == sizeof (CK_BBOOL) && *(CK_BBOOL *) &privateAttrib.front() != CK_TRUE)
+					continue;
+
+				GetObjectAttribute (slotId, dataHandle, CKA_MODULUS_BITS, privateAttrib);
+				if (privateAttrib.size() == sizeof (CK_ULONG)) {
+					CK_ULONG modulus = *(CK_ULONG *) &privateAttrib.front();
+					key.maxDecryptBufferSize = modulus / 8 - 11;
+					key.maxEncryptBufferSize = modulus / 8;
+				}
+
```

Comment:
This logic will fail for cards that contain ECC keys. Since the current implementation supports only RSA keys, the code should use CKA_KEY_TYPE to restrict logic to RSA keys only (CKK_RSA)